dropwizard-health/src/main/java/io/dropwizard/health/response/DetailedJsonHealthResponseProvider.java

```diff
@@ -0,0 +1,82 @@
+package io.dropwizard.health.response;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.dropwizard.health.HealthStateAggregator;
+import io.dropwizard.health.HealthStateView;
+import io.dropwizard.health.HealthStatusChecker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.ws.rs.core.MediaType;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class DetailedJsonHealthResponseProvider implements HealthResponseProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DetailedJsonHealthResponseProvider.class);
+
+    @Nonnull
+    private final HealthStatusChecker healthStatusChecker;
+    @Nonnull
+    private final HealthStateAggregator healthStateAggregator;
+    @Nonnull
+    private final ObjectMapper mapper;
+
+    public DetailedJsonHealthResponseProvider(@Nonnull final HealthStatusChecker healthStatusChecker,
+                                              @Nonnull final HealthStateAggregator healthStateAggregator,
+                                              @Nonnull final ObjectMapper mapper) {
+        this.healthStatusChecker = Objects.requireNonNull(healthStatusChecker);
+        this.healthStateAggregator = Objects.requireNonNull(healthStateAggregator);
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Nonnull
+    @Override
+    public HealthResponse fullHealthResponse(@Nullable String type) {
+        final Collection<HealthStateView> healthStateViews = healthStateAggregator.healthStateViews();
+        final String responseBody;
+        try {
+            responseBody = mapper.writeValueAsString(healthStateViews);
+        } catch (final Exception e) {
+            LOGGER.error("Failed to serialize health state views: {}", healthStateViews, e);
+            throw new RuntimeException(e);
+        }
+        final boolean healthy = healthStatusChecker.isHealthy(type);
+
+        return new HealthResponse(healthy, responseBody, MediaType.APPLICATION_JSON);
+    }
+
+    @Nonnull
+    @Override
+    public HealthResponse minimalHealthResponse(@Nullable String type) {
+        final boolean healthy = healthStatusChecker.isHealthy(type);
+
+        return new HealthResponse(healthy, null, MediaType.APPLICATION_JSON);
+    }
+
+    @Nonnull
+    @Override
+    public HealthResponse partialHealthResponse(@Nullable final String type, @Nonnull final Collection<String> names) {
+        final Collection<HealthStateView> healthStateViews = names.stream()
+            .map(healthStateAggregator::healthStateView)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .collect(Collectors.toList());
+        String responseBody = null;
+        try {
```

Comment:
It's not clear to me how invalid health check requests would be handled. For example if invalid values were provided for `type` or `names`. The same concern would apply to `fullHealthResponse()` above.

One idea that comes to mind is that `HealthResponse` would have 3 possible values for `healthy` instead of being a boolean:
* `healthy`
* `unhealthy`
* `invalid_request`

When performing health checks in general there will always be an issue of dealing with indeterminate results. It is probably bad if the health check is invalid due to bad params, but it doesn't mean that the actual service is unhealthy.

Reply:
So with respect to names, we will return an empty optional from `healthStateAggregator::healthStateView` for any name that doesn't exist. This makes it so we allow users to provide names in query params that don't necessarily map to a health check without blowing up (which allows us to handle backwards compatibility more gracefully for when a user adds or removes a check, and the query params don't reflect that change right away.

In terms of the type, that's a good question. We also take the approach of gracefully handling invalid or unrecognized type values. 

In `HealthCheckManager::isHealthy` you can see this. Essentially, if the string provided is `ALIVE` (case-insensitive), then we return the right value for a liveness check, otherwise we just default to a readiness check result.

If a user passed in `STARTUP` (which could correspond to the [relatively new Kubernetes Startup Probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes)), we would want users to start passing in that query param without blowing up. But we might default to still returning the readiness probe result until that code change rolled out.

It's kind of just the philosophy of backwards compatibility and lax validations we're taking with this default response provider, to make it as broadly useful as possible. The important thing is that we allow any user to provide an alternative implementation to this via the usual Dropwizard factory polymorphic config means if they want different behavior though!